axum/src/extract/state.rs

```diff
@@ -0,0 +1,206 @@
+/// let handler_with_state = handler.with_state(state);
+///
+/// # async {
+/// axum::Server::bind(&"0.0.0.0:3000".parse().unwrap())
+///     .serve(handler_with_state.into_make_service())
+///     .await
+///     .expect("server failed");
+/// # };
+/// ```
+///
+/// # Substates
+///
+/// [`State`] only allows a single state type but you can use [`From`] to extract "substates":
+///
+/// ```
+/// use axum::{Router, routing::get, extract::State};
+///
+/// // the application state
+/// #[derive(Clone)]
+/// struct AppState {
+///     // that holds some api specific state
+///     api_state: ApiState,
+/// }
+///
+/// // the api specific state
+/// #[derive(Clone)]
+/// struct ApiState {}
+///
+/// // support converting an `AppState` in an `ApiState`
+/// impl From<AppState> for ApiState {
+///     fn from(app_state: AppState) -> ApiState {
+///         app_state.api_state
+///     }
+/// }
+///
+/// let state = AppState {
+///     api_state: ApiState {},
+/// };
+///
+/// let app = Router::with_state(state)
+///     .route("/", get(handler))
+///     .route("/api/users", get(api_users));
+///
+/// async fn api_users(
+///     // access the api specific state
+///     State(api_state): State<ApiState>,
+/// ) {
+/// }
+///
+/// async fn handler(
+///     // we can still access to top level state
+///     State(state): State<AppState>,
+/// ) {
+/// }
+/// # let _: Router<AppState> = app;
+/// ```
+///
+/// # For library authors
+///
+/// If you're writing a library that has an extractor that needs state, this is the recommended way
+/// to do it:
+///
+/// ```rust
+/// use axum_core::extract::{FromRequest, RequestParts};
+/// use async_trait::async_trait;
+/// use std::convert::Infallible;
+///
+/// // the extractor your library provides
+/// struct MyLibraryExtractor;
+///
+/// #[async_trait]
+/// impl<B, S> FromRequest<B, S> for MyLibraryExtractor
+/// where
+///     B: Send,
+///     // keep `S` generic but require that it can produce a `MyLibraryState`
+///     // this means users will have to implement `From<UserState> for MyLibraryState`
+///     S: Into<MyLibraryState> + Clone + Send,
+/// {
+///     type Rejection = Infallible;
+///
+///     async fn from_request(req: &mut RequestParts<B, S>) -> Result<Self, Self::Rejection> {
+///         // get a `MyLibraryState` from the shared application state
+///         let state: MyLibraryState = req.state().clone().into();
+///
+///         // ...
+///         # todo!()
+///     }
+/// }
+///
+/// // the state your library needs
+/// struct MyLibraryState {
+///     // ...
+/// }
+/// ```
+///
+/// Note that you don't need to use the `State` extractor since you can access the state directly
+/// from [`RequestParts`].
+#[derive(Debug, Default, Clone, Copy)]
+pub struct State<S>(pub S);
+
+#[async_trait]
+impl<B, OuterState, InnerState> FromRequest<B, OuterState> for State<InnerState>
+where
+    B: Send,
+    OuterState: Clone + Into<InnerState> + Send,
```

Comment:
I wonder whether cloning the outer state to then access a single field could prove to be a problematic pessimization. The only alternative I can think of is using a custom trait instead of `Into<InnerState>`, something that has `Clone` as a supertrait and a `&self` conversion method, and blanket `impl<T: Clone> ThatTrait<T> for T` like `From` (but in this case mapping `&T -> T`.

Reply:
Yeah thats a good point. We could do something like this

```rust
pub trait FromRef<T>: Clone {
    fn from_ref(value: &T) -> Self;
}

impl<T> FromRef<T> for T
where
    T: Clone,
{
    fn from_ref(value: &T) -> Self {
        value.clone()
    }
}
```

Reply:
Slight nitpick: With a general name like `FromRef`, maybe the trait itself shouldn't have `Clone` as a supertrait, i.e. maybe it should only be a bound on the blanket impl.

Reply:
I made a separate PR for this https://github.com/tokio-rs/axum/pull/1268